### ISSUE TITLE
Allow newlines in font:write

### ISSUE THREAD

**Issue by Sanic** [dividuum/info-beamer#20]
I'am trying to display the contents of multiple textfiles, where each of them contains newlines.
Unfortunately, the newlines will not be rendered as expected with font:write. The newline will just be shown as a blank.

Is there any way to do this?

**Comment by weeheavy** [comment]
I'd love to have this too, for the same reason as @Sanic

**Comment by superlou** [comment]
I hacked something quickly for wrapping [when using non-monospace fonts](https://github.com/superlou/next-beamer/blob/master/text_util.lua).  It's pretty crude (doesn't understand hyphenation), but I think I can add newline support pretty easily.

**Comment by dividuum** [comment]
Text rendering of any non-trivial text is kind of a weak point right now. I'm not sure what the best approach to improve this without adding a ton of complexity. For full unicode support, something like [harfbuzz.org](http://harfbuzz.org) is probably the way to go. I'm not sure how one would solve hyphenation and other locale dependent problems.

**Comment by superlou** [comment]
I agree that some kind of text-shaping library would be a more universal solution.  Would full Cairo be overkill?  Making PNGs using Cairo/Pango, then loading them as image resources seems a little inefficient, but it might be good enough for some use cases.

**Comment by dividuum** [comment]
> Would full Cairo be overkill? 

Good question. I guess that would require a complete overhaul of the existing font rendering code. Right now (in the Pi version) I'm rendering glyphs into a dynamically growing texture ([Example](https://cdn.infobeamer.com/s/img/bench-font-atlas.png)), then render strings as vertex buffers. For performance I cache those, so they can get reused across frames.

If I understand the scope of pango correctly, you basically hand it a string and get layout information for each glyph based on various rules ([Fascinating!](https://github.com/GNOME/pango/blob/master/pango/break.c)). I'm not sure how all of that would work together. All while being completely backwards compatible with the existing text rendering output.

**Comment by dividuum** [comment]
If it's a variable you source from somewhere (like a JSON file), be sure that its content is UTF-8 encoded. If you're using a string literal, make sure your editor actually saves the file in UTF-8.

**Comment by jegue** [comment]
thanks, I will try. let you know!
